specification/index.bs

```diff
@@ -190,7 +199,70 @@ Issue(62): Specify localization handling.
 
 ### Content scripts
 
-#### Isolated worlds
+<dfn>Content scripts</dfn> represent a set of JS and CSS files that should be injected into pages loaded by the user agent.
+
+#### Key `matches`
+
+A [=list=] of [=match patterns=] that are used to decide where the content script runs. This key is required.
+
+#### Key `exclude_matches`
+
+A [=list=] of [=match patterns=] that should be used to exclude URLs from where the content script runs.
+
+#### Key `js`
+
+A [=list=] of file paths that should be injected as scripts.
+
+#### Key `css`
+
+A [=list=] of file paths that should be injected as stylesheets.
+
+#### Key `all_frames`
+
+If `all_frames` is true, the content script must be injected into subframes. Defaults to false.
+
+#### Key `match_about_blank`
+
+If this is `true`, the content script will also be injected into an additional user agent specified set of pages used to represent empty frames. This will only happen if the content script matches the page that embedded the frame. Defaults to `false`.
+
+#### Key `match_origin_as_fallback`
+
+Used to match frames with an opaque or otherwise missing origin. The origin to match against is determined in the following order of priority:
+
+1. If the frame has an [=opaque origin=], such as with a [=blob URLs=], use the non-opaque origin.
```

Comment:
@Rob--W @rdcronin Would you be able to take a look at this one and confirm if it is accurate? This was my best understanding based on bugs and documentation in the code.

Reply:
I think I'd rephrase this.

(The issue with the current language is that:
a] it doesn't specify what the "non-opaque" origin is or where it comes from
b] it doesn't always use the origin of the parent; it uses the initiator (or "precursor"))

If the URL of a document has a specified scheme**, the user agent will fall back to the origin of the initiator instead.  This is commonly, but not always, the parent or embedding frame.

** In chrome, these schemes are data:, about:, filesystem:, and blob:.  Is that the same in other browsers?

Reply:
@oliverdunk The semantics have extensively been discussed on Chromium's issue tracker where I and Devlin discussed the API design. If you're interested, the start of the discussion is at https://issues.chromium.org/issues/40443085#comment48. The design that is close to what we have now was sketched in https://issues.chromium.org/issues/40443085#comment61 , with the final name (`match_origin_as_fallback`) at https://issues.chromium.org/issues/40443085#comment67. Devlin summarized the discussion at https://issues.chromium.org/issues/40443085#comment71

Upon reviewing the proposed texts here, I think that there is some confusion on terminology. The current text mentions blob URLs as an opaque origin, but that is not the case.

Relevant to content script matching is the URL of the document (which can have an origin component) and the origin of the document (as a security principal). There may not always be an obvious relation between the two:

- URLs may have visible origin parts in it, such as http(s) and also `blob:` and (Chrome-only) `filesystem:` (e.g. `blob:https://example.com/UUID`).
- URLs may not have a visible origin in it (`about:blank` and `about:srcdoc`), but still have a non-opaque origin: commonly the opener of the frame or window is another http(s) URL. Or even any number of `about:blank`/`srcdoc` documents where the first was initially opened by a http(s) origin.
- The security principal of a document can be an [opaque origin](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque), even if the URL of that document looks like it has a non-opaque origin. This happens with `<iframe sandbox>` or `sandbox` directive in the `Content-Security-Policy`. A content script can use `window.origin` to see whether the origin is opaque, as it would serialize to `"null"`.
  - In case of opaque origins, there is almost always a non-opaque initiator that opened the frame. The term "precursor origin" is used here `<iframe sandbox="allow-scripts" src="https://example.com">`
 - The exception is when the initiator of the navigation does not have a non-opaque origin. For example, when the user navigates to a `data:-URL` or to `about:blank`. Since `data:`-URL
   - Chrome does currently not run content scripts in these documents.
   - Firefox currently allows content scripts with `matches` for all URLs AND `match_about_blank: true` to run scripts in top-level about:blank. This is not documented anywhere though.

Reply:
Thanks both, writing the description for these two keys has taken by far the most time in this PR. I've given it another attempt and would appreciate any feedback.

As a general note, concepts like the precursor origin and security principal don't appear to be defined in any other specifications. It seems like they are more informal terms used often in implementations and by implementors. With that in mind, I've tried to describe them as best as possible without talking about them by name.

A few additional notes:

- I've added an informal note to `match_about_blank` describing the Firefox behavior for top-level about:blank pages.
- I've added a note that the path must be a wildcard if `match_origin_as_fallback` is set. This is the behavior today in Chrome. Interestingly, we don't have any restrictions on `include_globs` or `exclude_globs`. This feels like an omission to me and I wonder if we should specify something.
- In Chrome, sandboxing doesn't seem to be relevant. We always apply these fallbacks, even if the parent is inaccessible to the child frame. With that in mind, I haven't mentioned it here.

Clearly there's a lot of detail here so please let me know if I've missed anything or it could be clearer.